torchvision/prototype/datasets/utils/_resource.py

```diff
@@ -141,8 +143,40 @@ def __init__(
         super().__init__(file_name=file_name or pathlib.Path(urlparse(url).path).name, **kwargs)
         self.url = url
         self.mirrors = mirrors
+        self._resolved = False
+
+    @property
+    def resolved(self) -> bool:
+        return self._resolved
+
+    def resolve(self) -> Optional[OnlineResource]:
+        redirect_url = _get_redirect_url(self.url)
+        if redirect_url == self.url:
+            self._resolved = True
+            return self
+
+        meta = {
+            attr.lstrip("_"): getattr(self, attr)
+            for attr in (
+                "file_name",
+                "sha256",
+                "_preprocess",
+                "_loader",
+            )
+        }
+
+        gdrive_id = _get_google_drive_file_id(redirect_url)
+        if gdrive_id:
+            return GDriveResource(gdrive_id, **meta)
+
+        http_resource = HttpResource(redirect_url, **meta)
+        http_resource._resolved = True
+        return http_resource
 
     def _download(self, root: pathlib.Path) -> None:
+        if not self.resolved:
+            return cast(OnlineResource, self.resolve())._download(root)
```

Comment:
Q: I'm seeing a few instances like this where we have to call `cast()` because the function we call returns an `Optional[...]`

Is there a less intrusive way of telling mypy that we *know* the output won't be `None`? Peherpas this is a configurable option?

In this very specific case though, it looks like the function never returns `None`, perhaps we could remove the `Optional` annotation?

Reply:
> Is there a less intrusive way of telling mypy that we _know_ the output won't be `None`? Peherpas this is a configurable option?

I don't think so as would go against the idea of type checking. If something is annotated as `Optional[Foo]` you should not use it as if it was an instance of `Foo`.

> In this very specific case though, it looks like the function never returns `None`, perhaps we could remove the `Optional` annotation?

Yes, that was a mistake on my side.